### PR TITLE
[decoupled-execution] add new message types: CommitProposal and Commit Decision.

### PR DESCRIPTION
**Pull request by yuxiamit**
Commit phase broadcasts commit proposals to seek agreement on the execution state.
Upon receiving a valid quorum of signatures on the same execution state, the nodes
will commit the execution result, and asynchronously broadcast a new commit
decision message containing the accumulated signatures to help slower nodes catch
up quickly. A node receiving a valid commit decision message does not have to wait
for commit proposals any more. It can directly commit the execution result.

Message handling is not included in this commit.

## Motivation

This is part of the decoupled execution project.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Unit tests will be added later after integration

## Related PRs

(If this PR adds or changes functionality, please take some time to update or suggest changes to the docs at https://developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

### CODE REVIEW

File: consensus/consensus-types/src/experimental/commit_decision.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    /// Return the LedgerInfo associated with this commit proposal
+    pub fn ledger_info(&self) -> &LedgerInfoWithSignatures {
+        &self.ledger_info
+    }
+
+    /// Verifies that the signatures carried in the message forms a valid quorum,
+    /// and then verifies the signature.
+    pub fn verify(&self, validator: &ValidatorVerifier) -> anyhow::Result<()> {
+        // We do not need to check the author because as long as the signature tree
+        // is valid, the message should be valid.
```
Comment (zekun000) [inline consensus/consensus-types/src/experimental/commit_decision.rs:57]: self.ledger_info.verify_signatures()

File: consensus/consensus-types/src/experimental/commit_proposal.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+use serde::{Deserialize, Serialize};
+use short_hex_str::AsShortHexStr;
+use diem_types::{
+    ledger_info::LedgerInfo, validator_signer::ValidatorSigner,
+    validator_verifier::ValidatorVerifier,
+};
+
+
+#[derive(Deserialize, Serialize, Clone, PartialEq, Eq)]
+pub struct CommitProposal {
```
Comment (zekun000) [inline consensus/consensus-types/src/experimental/commit_proposal.rs:17]: this is more like a Vote than a Proposal, probably CommitVote

File: consensus/src/network_interface.rs [hunk 3]
```diff
@@ -47,6 +47,13 @@ pub enum ConsensusMsg {
     /// VoteMsg is the struct that is ultimately sent by the voter in response for receiving a
     /// proposal.
     VoteMsg(Box<VoteMsg>),
+    /// CommitProposal is the struct that is sent by the validator after execution to propose
+    /// on the committed state hash root.
+    CommitProposalMsg(Box<CommitProposal>),
+    /// CommitDecision is the struct that is sent by the validator after collecting no fewer
+    /// than 2f + 1 signatures on the commit proposal. This part is not on the critical path, but
+    /// it can save slow machines to quickly confirm the execution result.
+    CommitDecisionMsg(Box<CommitDecision>),
```
Comment (zekun000) [inline consensus/src/network_interface.rs:56]: I'd prefer to embed this in SyncInfo, maybe we can leave CommitDecision for later and only add CommitProposal for now

File: consensus/consensus-types/src/experimental/commit_decision.rs (hunk 1 above)
Comment (yuxiamit) [reply consensus/consensus-types/src/experimental/commit_decision.rs:57]: Thanks

File: consensus/consensus-types/src/experimental/commit_proposal.rs (hunk 2 above)
Comment (yuxiamit) [reply consensus/consensus-types/src/experimental/commit_proposal.rs:17]: Resolved

File: consensus/src/network_interface.rs (hunk 3 above)
Comment (yuxiamit) [reply consensus/src/network_interface.rs:56]: Agree. We can make this change later

**Review by zekun000 — DISMISSED** [review]

**Pull request closed without merging**